### PR TITLE
[inputs.vsphere] Resolved issue 4790 (Resource whitelisting)

### PR DESCRIPTION
**Pull request by prydin**
### Required for all PRs:

- [x] Signed [CLA](https://influxdata.com/community/cla/).
- [x] Associated README.md updated.
- [x] Has appropriate unit tests.

Addresses issues #4789 and #4989 and allows users to select resources using inventory paths. Each resource type is given a *_include clause in the config file that looks like this:

```toml
vm_include = [ "*/vm/a_folder/a_subfolder/foobar*" ]
```

@glinton and @danielnelson This PR also has the changes provided in #5113. I'd love to get this in for 1.10 and it would probably make sense to merge this one and retire #5113 in that case.

### CODE REVIEW

**Comment by prydin** [comment]
@danielnelson @glinton Is the plan still to pull this in by 1.10? I haven't seen much movement on it.

**Comment by danielnelson** [comment]
I'd like to include it, can you rebase?

File: plugins/inputs/vsphere/README.md [hunk 1]
```diff
@@ -27,6 +27,7 @@ vm_metric_exclude = [ "*" ]
 
   ## VMs
   ## Typical VM metrics (if omitted or empty, all metrics are collected)
+  # vm_include = [ "/*/vm/**"] # Inventory path to VMs to collect (by default all are collected)
```
[the remark this answers is not in the document]
Comment (prydin) [reply plugins/inputs/vsphere/README.md]: I could, but the code would be a lot more complex and less efficient. Given an arbitrary path, it's hard to determine which type of object it's going to resolve to. For example "/datacenter1/**" is a valid path. You'd have to match every path against every object type, which would affect performance negatively. I suggest we keep it as it it.

File: Gopkg.lock [hunk 2]
```diff
@@ -1110,17 +1110,6 @@
   revision = "3617f28d167d448f93f282a867870f109516d2a5"
   version = "v0.19.0"
 
-[[projects]]
```
[the remark this answers is not in the document]
Comment (prydin) [reply Gopkg.lock]: Fixed!

File: plugins/inputs/vsphere/vsphere_test.go [hunk 3]
```diff
@@ -254,13 +339,17 @@ func TestTimeout(t *testing.T) {
...
 	defer s.Close()
 
-	var acc testutil.Accumulator
 	v := defaultVSphere()
+	var acc testutil.Accumulator
 	v.Vcenters = []string{s.URL.String()}
 	v.Timeout = internal.Duration{Duration: 1 * time.Nanosecond}
 	require.NoError(t, v.Start(nil)) // We're not using the Accumulator, so it can be nil.
 	defer v.Stop()
+<<<<<<< HEAD
```
[the remark this answers is not in the document]
Comment (prydin) [reply plugins/inputs/vsphere/vsphere_test.go]: Fixed!

File: plugins/inputs/vsphere/vsphere_test.go [hunk 4]
```diff
@@ -333,4 +525,8 @@ func TestAll(t *testing.T) {
 	defer v.Stop()
 	require.NoError(t, v.Gather(&acc))
 	require.Equal(t, 0, len(acc.Errors), fmt.Sprintf("Errors found: %s", acc.Errors))
+<<<<<<< HEAD
```
[the remark this answers is not in the document]
Comment (prydin) [reply plugins/inputs/vsphere/vsphere_test.go]: Fixed!

File: plugins/inputs/vsphere/README.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (prydin) [reply plugins/inputs/vsphere/README.md]: Fixed!

File: plugins/inputs/vsphere/panic_handler.go [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+
+import (
+	"errors"
+	"fmt"
+	"log"
+
+	"github.com/influxdata/telegraf"
+)
+
+func HandlePanicWithAcc(acc telegraf.Accumulator) {
```
[the remark this answers is not in the document]
Comment (prydin) [reply plugins/inputs/vsphere/panic_handler.go]: Fixed!

**Comment by prydin** [comment]
Fixed the merge issues, but circleci is failing at a step where it's trying to download some package. Don't think that's related to this PR. Maybe something intermittent in the test harness?

File: Gopkg.lock [hunk 6]
```diff
@@ -1111,17 +1111,17 @@
   version = "v0.19.0"
 
 [[projects]]
-  digest = "1:c3bdfb7e9b2a66bafbd47517a1a4e489706f75af37ad5bfb57621bf41c16b556"
```
[the remark this answers is not in the document]
Comment (prydin) [reply Gopkg.lock]: Fixed!

**Pull request merged**